### ISSUE TITLE
[Vulkan] numeric issue for quantized vision models

### ISSUE THREAD

**Issue by mariecwhite** [iree-org/iree#8731]
Many TFLite models are failing on vulkan with the error `failed to legalize operation 'arith.constant'`:

```
error: failed to legalize operation 'arith.constant'
note: called from
error: failed to run translation of source executable to target executable for backend #hal.executable.target<"vulkan", "vulkan-spirv-fb", {spv.target_env = #spv.target_env<#spv.vce<v1.3, [Shader, GroupNonUniform, GroupNonUniformVote, GroupNonUniformArithmetic, GroupNonUniformBallot, GroupNonUniformShuffle, GroupNonUniformShuffleRelative], [SPV_KHR_storage_buffer_storage_class]>, SwiftShader:CPU, {cooperative_matrix_properties_nv = [], max_compute_shared_memory_size = 16384 : i32, max_compute_workgroup_invocations = 128 : i32, max_compute_workgroup_size = dense<[128, 128, 64]> : vector<3xi32>, subgroup_size = 4 : i32}>}>
```

List of models that fail to compile:
```
mobilebert_tf2_quant_test.py
bird_classifier_test.py
efficientnet_lite0_int8_test.py
image_stylization_test.py
inception_v4_uint8_test.py
lightning_i8_test.py
magenta_test.py
mobilenet_ssd_quant_test.py
mobilenet_ssd_quant_test.py
mobilenet_v1_uint8_test.py
mobilenet_v2_int8_test.py
mobilenet_v2_uint8_test.py
mobilenet_v3-large_uint8_test.py
mobilenet_v35_int8_test.py
person_detect_test.py
resnet_50_int8_test.py
ssd_mobilenet_v1_uint8_test.py
visual_wake_words_i8_test.py
```

MLIR files for mobilebert_tf2_quant_test:
[tflite.mlir.zip](https://github.com/google/iree/files/8399857/tflite.mlir.zip)
[tosa.mlir.zip](https://github.com/google/iree/files/8399858/tosa.mlir.zip)

**Comment by benvanik** [comment]
Do you know what the arith.constant is? (would help understand if it's a data type issue)

**Comment by rsuderman** [comment]
I think its related to certain types not lowering to spirv - my guess would be that TOSA import does not handle the sample `i64` to `i32` narrowing.

**Comment by benvanik** [comment]
If so that's great - #8738 will add a new `-iree-flow-demote-i64-to-i32` flag that can be used for models that have i64 that should help!

**Comment by benvanik** [comment]
#8738 may not fix tosa due to #8745 - it may need a corresponding `-iree-tosa-demote-i64-to-i32` on the tosa import path to the `-iree-mhlo-demote-i64-to-i32` on the mhlo path until it's fixed.

**Comment by rsuderman** [comment]
So what is weirder is the `i64` values seem to be constructed later in the lowering process. Working to understand why at the moment.

**Comment by rsuderman** [comment]
Alright, figured it out. The problem is the conversion for `tosa.apply_scale` generates the `i64` operations to handle the rescaling behavior. This creates `i64` types that the gpu doesn't like. This is not supposed to happen until further in the pipeline but was accidentally converted after the TosaToLinalg. I am going to tweak the TosaToStandard pass to optionally disable the `tosa.apply_rescale` lowering so that it is optionally disabled.

**Comment by rsuderman** [comment]
This should handle the fix once it lands and integrates: https://reviews.llvm.org/D122948

**Comment by antiagainst** [comment]
I'm not seeing the compilation failures anymore with the latest tip of the tree. So I assume it was fixed by Rob. @mariecwhite could you double check and close the issue if so? Thanks!

**Comment by mariecwhite** [comment]
These models are now compiling but the numerical differences are large

```
> python tflitehub/mobilebert_tf2_quant_test.py --config=vulkan
...
I0503 12:13:31.038540 140054783264576 test_util.py:72] Max error (0): 44.021111
I0503 12:13:31.038809 140054783264576 test_util.py:72] Max error (1): 51.602524
...
```
```
> python tflitehub/efficientnet_lite0_int8_test.py --config=vulkan
...
I0503 12:14:50.095983 139686240139072 test_util.py:72] Max error (0): 249.000000
...
```
```
> python tflitehub/image_stylization_test.py --config=vulkan
...
I0503 12:15:54.304497 139976542127936 test_util.py:72] Max error (0): 1.642846
...
```

**Comment by mariecwhite** [comment]
Reassigning to Lei to take a further look.

**Comment by antiagainst** [comment]
The issue is still `tosa.apply_scale` lowering generating `i64` `arith` ops, though now in `ConvertToSPIRVPass` itself. (The TOSA lowering should be moved earlier in the pipeline, but that's a separate issue.) There are bugs regarding `i64` `arith` ops in SPIR-V conversion right now if the target does not natively support `i64`. I expect this to be fixed once we have https://github.com/iree-org/iree/issues/8702.

**Comment by benvanik** [comment]
I don't know if "fixed" so much as worked-around - having i64 emulation on hot paths isn't great :)
If we do start putting i64 emulation on hot paths we should make sure to track that - it's the kind of death-by-a-thousand-papercuts work that is hard to track down later and lowers the performance ceiling. Do we know if any other tosa-to-vulkan target (are there any?) does this?

**Comment by antiagainst** [comment]
Yup, I see your points and agreed. :) It's best to avoid the 64-bit calculation entirely if possible. That would save us from all these pain and performance penalty. I'm not super familiar with all the numeric nuances  behind `tosa.apply_scale`; just from my quick reading over the `tosa.apply_scale` spec, it might be hard given the whole definition is tied to i64 and "All arithmetic is designed so that it does not overflow a 64-bit accumulator and that the final result fits in 32 bits". It would be great if @rsuderman can weigh in on this.

Suppose that we are really stuck with i64 arithmetics here then there is no out other than emulate it, unless the target supports i64 natively. (I verified that if we compile these models towards `ampere-unknown-unknown` they gives the correct results.) The rescaling [implementation](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/TosaToArith/TosaToArith.cpp#L61-L118) requires mainly add, mul, and shift. Add and shift is fine. Mul is kinda expensive, though it's only one instance. So hopefully for this specific `tosa.apply_scale` case we won't see too much of a problem. Emulating a whole program with i64 widely used (just because of the frontends by default emits i64, not really needing i64 for numerics) is certainly something we should be very cautious.

**Comment by rsuderman** [comment]
@antiagainst 

Good news: we have the i32 only version of the rescale lowering, which is pretty trivial to enable. There is an optional boolean for whether to include the 32-bit-only replacement pattern:
https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/SPIRV/ConvertToSPIRVPass.cpp#L347

Bad news: our `tosa.mul` test fails when it is enabled. We know that the rescale lowering is not the issue as we can enable it on the CPU / VMVX backends and they don't fail. So something must be going wrong further on in the SPIRV lowering.

**Comment by kuhar** [comment]
I tried the first model from the list with I64 emulation enabled. It compiles and the error is much smaller than the initial (51.6 -> 7.5), but still it doesn't pass the `np.isclose` check:

```
I1116 17:29:14.608334 140613764630336 test_util.py:82] Setting up iree runtime
I1116 17:29:45.152059 140613764630336 test_util.py:131] Setting up test inputs
I1116 17:29:45.152978 140613764630336 mobilebert_tf2_quant_test.py:18]  [  1 384], int32
I1116 17:29:45.153103 140613764630336 mobilebert_tf2_quant_test.py:18]  [  1 384], int32
I1116 17:29:45.153243 140613764630336 mobilebert_tf2_quant_test.py:18]  [  1 384], int32
I1116 17:29:45.153416 140613764630336 test_util.py:134] Invoking TFLite
I1116 17:29:45.461065 140613764630336 test_util.py:96] Invocation time: 0.3074 seconds
I1116 17:29:45.461337 140613764630336 test_util.py:137] Invoke IREE
I1116 17:29:48.449509 140613764630336 test_util.py:111] Invocation time: 2.9878 seconds
I1116 17:29:48.450227 140613764630336 test_util.py:72] Max error (0): 6.114044
I1116 17:29:48.450379 140613764630336 test_util.py:72] Max error (1): 7.581413
[  FAILED  ] MobileBertTest.test_compile_tflite
======================================================================
FAIL: test_compile_tflite (__main__.MobileBertTest)
MobileBertTest.test_compile_tflite
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/kubak/iree/iree-samples/tflitehub/mobilebert_tf2_quant_test.py", line 36, in test_compile_tflite
    self.compile_and_execute()
  File "/usr/local/google/home/kubak/iree/iree-samples/tflitehub/test_util.py", line 147, in compile_and_execute
    self.compare_results(iree_results, tflite_results, self.output_details)
  File "/usr/local/google/home/kubak/iree/iree-samples/tflitehub/mobilebert_tf2_quant_test.py", line 33, in compare_results
    self.assertTrue(np.isclose(iree_results[1], tflite_results[1], atol=7.0).all())
AssertionError: False is not true
```

Do you know how the error constant (7.0) was picked and if there could be another sources of error outside of i64 truncation? @mariecwhite @rsuderman @antiagainst

**Comment by mariecwhite** [comment]
This was chosen after running MobileBert quantized on a large dataset (Squad v1.1) on x86. We used MLPerf to verify that the final accuracy number was within 0.1% of TFLite. So even though the difference here was large on a single example, running it on thousands of examples resulted in the same aggregated accuracy metric as TFLite.

If the difference is 7.5, it's likely that the numeric issues are fixed. If you would like to double-check and re-run that accuracy test, instructions are here: https://github.com/mariecwhite/mobile_app_open/tree/linux_workspace/mobile_back_iree.

**Comment by allieculp** [comment]
@mariecwhite still open?

**Comment by mariecwhite** [comment]
I believe @kuhar was looking into this

**Comment by allieculp** [comment]
@kuhar Can you please help update this stale bug? Please deprioritize/close as needed.

**Comment by kuhar** [comment]
This needs another round of triage to decide whether the remaining differences are acceptable or not, like @mariecwhite suggested  in https://github.com/openxla/iree/issues/8731#issuecomment-1319273795.

Not sure what you mean by 'deprioritize' -- I don't see any priority assigned to this bug.

**Comment by allieculp** [comment]
This is labeled as a P1 currently, we can leave as is until more triage has been done!

**Comment by kuhar** [comment]
Ahhh, I found it under the project drop down menu:
<img width="323" alt="image" src="https://user-images.githubusercontent.com/4612584/230480853-8e3fadb5-263b-4242-b3d3-0d145c0e8b14.png">

I changed it to P2. Thanks, @allieculp!

**Comment by antiagainst** [comment]
Closing this for now -- no action planned for now
